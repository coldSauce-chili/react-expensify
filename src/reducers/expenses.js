
const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]

        case 'REMOVE_EXPENSE':
            const filtered = state.filter((expense) => expense.id !== action.id)
            return filtered

        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })

        default:
            return state
    }
}

export default expensesReducer