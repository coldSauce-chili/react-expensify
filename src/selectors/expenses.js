import moment from 'moment'

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase())


        //before momentjs
        //comparing only by raw epoch time
        //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate
        //const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
        // else {
        //     return 0
        // }
    })

}

