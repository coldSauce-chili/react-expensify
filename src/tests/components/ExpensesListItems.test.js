import React from 'react'
import { shallow } from 'enzyme'
import ExpenseListItems from '../../components/ExpenseLIstItem'
import expenses from '../fixtures/expenses'

test('this should render an expense item', () => {
    const wrapper = shallow(<ExpenseListItems {...expenses[1]} />)

    expect(wrapper).toMatchSnapshot()
})