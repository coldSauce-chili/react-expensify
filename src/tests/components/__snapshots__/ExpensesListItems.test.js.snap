// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`this should render an expense item 1`] = `
<div>
  <Link
    replace={false}
    to="edit/2"
  >
    <h4>
      description: 
      rent
    </h4>
  </Link>
  <h5>
    amount: 
    109500
  </h5>
  <h6>
    date: 
    -345600000
  </h6>
</div>
`;
